### PR TITLE
DefaultHttp2Connection.close Reentrant Modification

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The DefaultHttp2Conneciton.close method accounts for active streams being iterated and attempts to avoid reentrant modifications of the underlying stream map by using iterators to remove from the stream map. However there are a few issues:

- While iterating over the stream map we don't prevent iterations over the active stream collection
- Removing a single stream may actually remove > 1 streams due to closed non-leaf streams being preserved in the priority tree which may result in NPE

Preserving closed non-leaf streams in the priority tree is no longer necessary with our current allocation algorithms, and so this feature (and related complexity) can be removed.

Modifications:
- DefaultHttp2Connection.close should prevent others from iterating over the active streams and reentrant modification scenarios which may result from this
- DefaultHttp2Connection should not keep closed stream in the priority tree
  - Remove all associated code in DefaultHttp2RemoteFlowController which accounts for this case including the ReducedState object
  - This includes fixing writability changes which depended on ReducedState
- Update unit tests

Result:
Fixes netty#5198

### CODE REVIEW

**Comment by Scottmitch** [comment]
@nmittler - PTAL

@ylgrgyq @jchambers - please verify.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 1]
```diff
@@ -218,8 +224,8 @@ public void goAwayReceived(final int lastKnownStream, long errorCode, ByteBuf de
         for (int i = 0; i < listeners.size(); ++i) {
             try {
                 listeners.get(i).onGoAwayReceived(lastKnownStream, errorCode, debugData);
-            } catch (RuntimeException e) {
-                logger.error("Caught RuntimeException from listener onGoAwayReceived.", e);
+            } catch (Throwable cause) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Why this change? Do we really want to silence errors?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 2]
```diff
@@ -249,8 +255,8 @@ public void goAwaySent(final int lastKnownStream, long errorCode, ByteBuf debugD
         for (int i = 0; i < listeners.size(); ++i) {
             try {
                 listeners.get(i).onGoAwaySent(lastKnownStream, errorCode, debugData);
-            } catch (RuntimeException e) {
-                logger.error("Caught RuntimeException from listener onGoAwaySent.", e);
+            } catch (Throwable cause) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Same comment.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 3]
```diff
@@ -369,7 +371,6 @@ final DefaultPropertyKey verifyKey(PropertyKey key) {
         private short weight = DEFAULT_PRIORITY_WEIGHT;
         private DefaultStream parent;
         private IntObjectMap<DefaultStream> children = IntCollections.emptyMap();
-        private int prioritizableForTree = 1;
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Remind me what this was for again ... did a previous incarnation of the priority algorithm need it?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 4]
```diff
@@ -679,29 +617,18 @@ final void takeChild(DefaultStream child, boolean exclusive, List<ParentChangedE
...
-                        // concrete cases known but is risky because it could invalidate the data structure.
-                        // 2. We are notifying listeners of the removal while the tree is in flux. Currently the
-                        // codec listeners make no assumptions about priority tree structure when being notified.
-                        removeStream(oldParent, null);
-                    }
+                // Note that the removal operation may not be successful and may return null. This is because when an
+                // exclusive dependency is processed the children are removed in a previous recursive call but the
+                // child's parent link is updated here.
+                if (oldParent != null) {
+                    oldParent.children.remove(child.id());
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: It's been a while since I looked closely at this code, so bare with me...

If this is being done by a tree restructure operation, can we get into trouble by removing immediately?

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java [hunk 5]
```diff
@@ -696,11 +697,6 @@ public void weightChangeWithNoTreeChangeShouldNotifyListeners() throws Http2Exce
                 any(Http2Stream.class));
         verify(clientListener, never()).onPriorityTreeParentChanged(any(Http2Stream.class),
                 any(Http2Stream.class));
-        assertEquals(5, client.connectionStream().prioritizableForTree());
```
Comment (nmittler) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java]: Seems like we're effectively removing a lot of testing.  Any thoughts on how we might add more verification to make up for the loss of coverage here?

**Comment by nmittler** [comment]
@Scottmitch took a first pass ... wow, things got a lot simpler! :)

**Comment by ylgrgyq** [comment]
Already put this fix to usage. But I need some time to check if the leak is gone. I'll come back to give my result as soon as possible. Thanks a lot.

**Comment by ylgrgyq** [comment]
Very sad to tell that the `NullPointerException` is still there:
```
[WARN] [nioEventLoopGroup-2-38] 05-05 14:17:26,050 [ApnsClientHandler] - APNs client pipeline caught an exception.
java.lang.NullPointerException
        at io.netty.handler.codec.http2.DefaultHttp2Connection.close(DefaultHttp2Connection.java:143)
        at io.netty.handler.codec.http2.Http2ConnectionHandler$BaseDecoder.channelInactive(Http2ConnectionHandler.java:179)
        at io.netty.handler.codec.http2.Http2ConnectionHandler.channelInactive(Http2ConnectionHandler.java:370)
        at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelInactiveNow(ChannelHandlerInvokerUtil.java:56)
        at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelInactive(DefaultChannelHandlerInvoker.java:93)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:342)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:124)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
        at io.netty.handler.timeout.IdleStateHandler.channelInactive(IdleStateHandler.java:257)
        at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelInactiveNow(ChannelHandlerInvokerUtil.java:56)
        at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelInactive(DefaultChannelHandlerInvoker.java:93)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:342)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:124)
        at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:344)
        at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:309)
        at io.netty.handler.ssl.SslHandler.channelInactive(SslHandler.java:727)
        at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelInactiveNow(ChannelHandlerInvokerUtil.java:56)
        at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelInactive(DefaultChannelHandlerInvoker.java:93)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:342)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:124)
        at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:1060)
        at io.netty.channel.AbstractChannel$AbstractUnsafe$7.run(AbstractChannel.java:747)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:339)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:374)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:742)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:145)
        at java.lang.Thread.run(Thread.java:745)
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 6]
```diff
@@ -1249,6 +1154,27 @@ void removeFromActiveStreams(DefaultStream stream, Iterator<?> itr) {
         boolean allowModifications() {
             return pendingIterations == 0;
```
Comment (ylgrgyq) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Does `pendingIterations` need to be marked as `volatile` ? In case of modified by several threads concurrently?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 7]
```diff
@@ -131,18 +131,24 @@ final boolean isClosed() {
             promise.trySuccess(null);
             return promise;
         }
+
         Iterator<PrimitiveEntry<Http2Stream>> itr = streamMap.entries().iterator();
         // We must take care while iterating the streamMap as to not modify while iterating in case there are other code
         // paths iterating over the active streams.
         if (activeStreams.allowModifications()) {
```
Comment (sunng87) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: If we were to avoid **concurrent** modification, why we don't use AtomicInteger here to track concurrent modification?
Or perhaps we were just to avoid modification in recursive calls?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 6 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Perhaps my usage of the term "concurrent modification" is misleading. The codec-http2 is not thread safe and what we are really protecting from is a re-entry condition on the same thread which may modify the collection we are iterating. This can happen for example because we call out to user code while iterating.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 7 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Perhaps "concurrent" was a poor choice of words in this context... see https://github.com/netty/netty/pull/5205#discussion_r62297682

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 6 above)
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Yep better update comment and use 'reentrance'

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @nmittler - For now yes. If we let exceptions propagate from user code our internal state may become corrupted if we have not completely updated it. I submitted a PR a while ago to try and allow exceptions to propagate but it needed more work and I have not got around to updating it (I will likely wait until childchannels comes to be).

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 3 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: this was the count of how many nodes rooted at the current node are not closed, and in effect are allowed to be in the priority tree by their own right. Previously a closed node may remain in priority tree if it had decedents that were not closed. The last priority algorithm didn't necessarily rely on this but its a feature that would complicate this fix and I don't believe it is necessary.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java (hunk 5 above)
Comment (Scottmitch) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java]: @nmittler - the testing that was removed was to support the `prioritizableForTree` feature which was removed in this PR. I don't think this feature is necessary and it would complicate the task of removing all streams while iterating over the stream map.

**Comment by Scottmitch** [comment]
@ylgrgyq - I would like to add a commit to this which includes some code which will help collect debug information when this condition occurs ... would you mind running it?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 6 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @normanmaurer - done

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 4 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: This is part of the tree restructure operation and is not removing from any other collection (e.g. active stream or stream map). This is just a subset of what was done above ... what was done above was necessary because a node that was closed could be restructured to be a leaf node ... its `prioritizableForTree` go to 0 ... and then we should remove it (closed nodes should not exist as leaf nodes). That operation was a bit more "risky" as indicated in the comments above ... but didn't lead to any concrete issues.

I think we are OK but did you have a specific case in mind?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 8]
```diff
@@ -277,18 +289,17 @@ private boolean isStreamMapEmpty() {
...
+     * @param stream the stream to remove.
+     * @param itr an iterator that may be pointing to the stream during iteration and {@link Iterator#remove()} will be
+     * used if non-{@code null}.
      */
     void removeStream(DefaultStream stream, Iterator<?> itr) {
-        // [1] Check if this stream can be removed because it has no prioritizable descendants.
         if (stream.parent().removeChild(stream)) {
             if (itr == null) {
+                if (isClosing) {
+                    logger.error("We are removing from the stream map while iterating!", new Throwable());
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @ylgrgyq - this is the log line I hope to see and would like you to report back if it does occur .. if it does not occur I would also like to know.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 9]
```diff
@@ -794,48 +702,58 @@ protected void initialWindowSize(int newWindowSize) throws Http2Exception {
...
         }
 
-        private void checkConnectionThenStreamWritabilityChanged(AbstractState state) throws Http2Exception {
+        private void checkConnectionThenStreamWritabilityChanged(FlowState state) throws Http2Exception {
             // It is possible that the connection window and/or the individual stream writability could change.
-            if (isWritableConnection() != connectionState.markWritability()) {
+            if (isWritableConnection() != connectionState.markedWritability()) {
                 checkAllWritabilityChanged();
-            } else if (isWritable(state) != state.markWritability()) {
+            } else if (isWritable(state) != state.markedWritability()) {
```
Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: MAJOR Prefix this call to "isWritable" with "super.". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2388)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 10]
```diff
@@ -657,21 +552,30 @@ public void write(Http2Stream stream, int numBytes) {
...
-        public void channelWritabilityChange() throws Http2Exception { }
+        void channelWritabilityChange() throws Http2Exception { }
 
         /**
-         * Called when the state is cancelled outside of a write operation.
+         * Called when the state is cancelled.
          * @param state the state that was cancelled.
          */
-        public void stateCancelled(AbstractState state) { }
+        void stateCancelled(FlowState state) { }
```
Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: MAJOR Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1186)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 11]
```diff
@@ -282,43 +256,83 @@ public void writePendingBytes() throws Http2Exception {
...
+         */
         private boolean writing;
-        // Set to true if cancel() was called.
+        /**
+         * Set to true if cancel() was called.
+         */
         private boolean cancelled;
+        private BooleanSupplier isWritableSupplier = new BooleanSupplier() {
+            @Override
+            public boolean get() throws Exception {
```
Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: MAJOR A getX() method which returns a boolean should be named isX() [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3ABooleanGetMethodName)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 12]
```diff
@@ -759,8 +663,8 @@ protected final boolean isWritableConnection() {
         private final Http2StreamVisitor checkStreamWritabilityVisitor = new Http2StreamVisitor() {
             @Override
             public boolean visit(Http2Stream stream) throws Http2Exception {
-                AbstractState state = state(stream);
-                if (isWritable(state) != state.markWritability()) {
+                FlowState state = state(stream);
+                if (isWritable(state) != state.markedWritability()) {
```
Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: MAJOR Prefix this call to "isWritable" with "super.". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2388)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 13]
```diff
@@ -657,21 +552,30 @@ public void write(Http2Stream stream, int numBytes) {
          * Called when the writability of the underlying channel changes.
          * @throws Http2Exception If a write occurs and an exception happens in the write operation.
          */
-        public void channelWritabilityChange() throws Http2Exception { }
+        void channelWritabilityChange() throws Http2Exception { }
```
Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: MAJOR Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1186)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java [hunk 14]
```diff
@@ -794,48 +702,58 @@ protected void initialWindowSize(int newWindowSize) throws Http2Exception {
...
+        void channelWritabilityChange() throws Http2Exception {
+            if (connectionState.markedWritability() != isChannelWritable()) {
                 checkAllWritabilityChanged();
             }
         }
 
-        private void notifyWritabilityChanged(AbstractState state) {
-            state.markWritability(!state.markWritability());
+        private void checkStateWritability(FlowState state) throws Http2Exception {
+            if (isWritable(state) != state.markedWritability()) {
```
Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: MAJOR Prefix this call to "isWritable" with "super.". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2388)

File: common/src/main/java/io/netty/util/BooleanSupplier.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+     * @throws Exception If an exception occurs.
+     */
+    boolean get() throws Exception;
+
+    /**
+     * A supplier which always returns {@code false} and never throws.
+     */
+    BooleanSupplier FALSE_SUPPLIER = new BooleanSupplier() {
+        @Override
+        public boolean get() {
```
Comment (netkins) [inline common/src/main/java/io/netty/util/BooleanSupplier.java]: MAJOR A getX() method which returns a boolean should be named isX() [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3ABooleanGetMethodName)

File: common/src/main/java/io/netty/util/BooleanSupplier.java [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * Represents a supplier of {@code boolean}-valued results.
+ */
+public interface BooleanSupplier {
+    /**
+     * Gets a boolean value.
+     * @return a boolean value.
+     * @throws Exception If an exception occurs.
+     */
+    boolean get() throws Exception;
```
Comment (netkins) [inline common/src/main/java/io/netty/util/BooleanSupplier.java]: MAJOR A getX() method which returns a boolean should be named isX() [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3ABooleanGetMethodName)

File: common/src/main/java/io/netty/util/BooleanSupplier.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+            return false;
+        }
+    };
+
+    /**
+     * A supplier which always returns {@code true} and never throws.
+     */
+    BooleanSupplier TRUE_SUPPLIER = new BooleanSupplier() {
+        @Override
+        public boolean get() {
```
Comment (netkins) [inline common/src/main/java/io/netty/util/BooleanSupplier.java]: MAJOR A getX() method which returns a boolean should be named isX() [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3ABooleanGetMethodName)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java (hunk 9 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: the outer class has a method named isWritable but the signature is different ... so not sure this will confuse anyone...

File: common/src/main/java/io/netty/util/BooleanSupplier.java (hunk 15 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/BooleanSupplier.java]: ignored .. the goal is to emulate `java.util.Supplier`

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java (hunk 11 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java]: ignored .. the goal is to emulate `java.util.Supplier`

**Comment by ylgrgyq** [comment]
@Scottmitch Got it. I'll come back latter.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 8 above)
Comment (ylgrgyq) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @Scottmitch - Sorry, I confirm that I got the same NPE and leak report several times after using this new commit but haven't saw this log yet.

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @ylgrgyq - thanks for reporting back. Humm this is the only spot where we remove from the map (besides with the iterator below) ... Let me think on it. Do have have a better ideas for a reproducer? Are you unregistering channels are accessing any codec-http2 APIs from outside the eventloop?

Comment (ylgrgyq) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @Scottmitch - Sorry for taking so long to response. I've been trying all day long to find a good way to reproduce this problem but still haven't got any progress yet. And for the second question, I think we don't accessing codec-http2 APIs from outside of the eventloop. All the client code which interacting with codec-http2's APIs is [this class](https://github.com/relayrides/pushy/blob/master/src/main/java/com/relayrides/pushy/apns/ApnsClientHandler.java). 

But I do have got something new other than reproducing the NPE. I change [`streamMap` in `DefaultHttp2Connection`](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L69) from `IntObjectHashMap` to `java.util.HashMap` (the declaration is `final Map<Integer, Http2Stream> streamMap = new HashMap<Integer, Http2Stream>()` ) I find that the leak and NPE is disappeared. Or at least for four hours I haven't seen it comparing to using `IntObjectHashMap` that NPE always jump out less than an hour in daytime under average load ( I'm at GMT+8 ). But I still need some time to make sure that the NPE is really gone. 

Now I'm trying to do some test for `IntObjectHashMap` to see weather this defect is in `IntObjectHashMap`.

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: > I change streamMap in DefaultHttp2Connection from IntObjectHashMap to java.util.HashMap (the declaration is final Map<Integer, Http2Stream> streamMap = new HashMap<Integer, Http2Stream>() ) I find that the leak and NPE is disappeared.

@ylgrgyq - Interesting ... I'll take a look at the `IntObjectHashMap` code

Comment (ylgrgyq) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Thank you very much. And I'll keep working on it too.
By the way, nearly 18 hours not seen any NPE or leak reports. I think they really gone.

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @ylgrgyq - I found an issue in `IntObjectHashMap` related to the iteration. I will submit a PR soon for you to verify. Thanks for your hard work :)

Comment (ylgrgyq) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: OK, no problem. It's my pleasure.

**Comment by Scottmitch** [comment]
@nmittler - PTAL (comments/questions answered)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 4 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: no, I think it's fine.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 18]
```diff
@@ -131,18 +131,24 @@ final boolean isClosed() {
...
         // paths iterating over the active streams.
         if (activeStreams.allowModifications()) {
-            while (itr.hasNext()) {
-                DefaultStream stream = (DefaultStream) itr.next().value();
-                if (stream.id() != CONNECTION_STREAM_ID) {
-                    // If modifications of the activeStream map is allowed, then a stream close operation will also
-                    // modify the streamMap. Pass the iterator in so that remove will be called to prevent concurrent
-                    // modification exceptions.
-                    stream.close(itr);
+            activeStreams.incrementPendingIterations();
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: @Scottmitch can we just call `forEachActiveStream` instead? I assume we can just clear the stream map after we're done?

**Comment by nmittler** [comment]
@Scottmitch I took another pass ... LGTM. Just one comment.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 18 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: > can we just call forEachActiveStream instead

Unfortunately not. The original motivation for this method was because we didn't close *all* the streams. There can be streams that are not active and thus not in the `activeStreams` collection that still must be closed.

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Ah right ... makes sense. Thanks!

**Comment by nmittler** [comment]
@Scottmitch LGTM

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (d580245)

@ylgrgyq - we can address the remainder of the issue on the next PR i open.

**Comment by ylgrgyq** [comment]
@Scottmitch - 👍  Great work!
